.github/workflows/main.yml

```diff
@@ -811,6 +811,33 @@ jobs:
         path: target/wasm32-unknown-unknown/release/wasi_snapshot_preview1.*.wasm
 
 
+    # common logic to cancel the entire run if this job fails
+    - run: gh run cancel ${{ github.run_id }}
+      if: failure() && github.event_name != 'pull_request'
+      env:
+        GH_TOKEN: ${{ github.token }}
+  
+  build-preview1-component-adapter-provider:
+    name: Build wasi-preview1-component-adapter-provider
+    needs: build-preview1-component-adapter
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        submodules: true
+    - uses: ./.github/actions/install-rust
+    - run: rustup component add rustfmt clippy
+
+    - uses: actions/download-artifact@v4
+      with:
+        name: bins-wasi-preview1-component-adapter
+        path: crates/wasi-preview1-component-adapter/provider/artefacts
+    
+    - run: ./ci/build-wasi-preview1-component-adapter-provider.sh
+      env:
+        CHECK: 1
```

Comment:
Could these steps be pulled into a composite action as well? That way this could be shared with the publish step, notably the configuration of the `download-artifact` step. Also I think it's ok to remove the `CHECK` variable and run the same checks on publication since it's a quick-to-build crate and we can throw some more tools like clippy on the publish step.

In theory that means that the publish step is:

1. Run the composite action to figure out the run-id 
2. Run the composite action to setup the artifacts and perform checks
3. Run a publish

That feels like a good state to me -- aka I like this refactoring 👍

Reply:
So this would be two new composite actions?

Reply:
Yeah, one for acquiring the run-id matching `github.sha` used during publish-artifacts and publish-to-cratesio. Another for verifying the adapter shared between main.yml and publish-to-cratesio